### PR TITLE
Code refactoring and first version of feedback implemented

### PR DESCRIPTION
**Pull request by kellhogs**
I've sepperated actions and feedback into their own files, and did some refactoring and updating of the code based on what i saw on some newer modules. i've also added feedback support for changing button styles when an input is in program or preview

### CODE REVIEW

File: analogway_livecore.js [hunk 1]
```diff
@@ -1,743 +1,564 @@
...
-			if (line.match(/E\d{2}/)) {
-				switch (parseInt(line.match(/E(\d{2})/)[1])) {
-					case 10: self.log('error',"Received command name error from "+ self.config.label +": "+ line); break;
-					case 11: self.log('error',"Received index value out of range error from "+ self.config.label +": "+ line); break;
-					case 12: self.log('error',"Received index count (too few or too much) error from "+ self.config.label +": "+ line); break;
-					case 13: self.log('error',"Received value out of range error from "+ self.config.label +": "+ line); break;
-					default: self.log('error',"Received unspecified error from Livecore "+ self.config.label +": "+ line);
+				if (line.match(/TAopr\d,\d/)) {
+					//Program Tally Information
+					let prearr = line.split(',')
```
Comment (dnmeid) [inline analogway_livecore.js:293]: prearr is never reasigned, use const

File: analogway_livecore.js [hunk 2]
```diff
@@ -1,743 +1,564 @@
...
+					let input = [Number(prearr[0].replace('TAopr', '')), prearr[1]]
```
Comment (dnmeid) [inline analogway_livecore.js:294]: input is the same as prearr, can be omitted. Just do replacement of element prearr[0]

File: analogway_livecore.js [hunk 3]
```diff
@@ -1,743 +1,564 @@
...
+					if (input[1] === '1') {
+						tallyPGM.push(input[0])
```
Comment (dnmeid) [inline analogway_livecore.js:296]: Number may be pushed multiple times, e.g. if a tally of command gets not thru. Then the feedback is broken.
Better implementation would be `tallyPGM[input[0]] = input[1]`. Then you have the state of each input in the corresponding index. Reading out would give 0 or 1 or undefined if the tally haven't been set.

File: analogway_livecore.js [hunk 4]
```diff
@@ -1,743 +1,564 @@
...
-					id: 'command',
-					default: '',
-					tooltip: "Enter any command you like in plain ASCII. Beware of correct syntax, you mustn't enter the linebreak at the end of the command.",
-					regex: '/^[\\w,]+$/i'
-			}]}
-	});
-}
+	feedback(feedback) {
+		if (feedback.type === 'input_active') {
+			if (tallyPGM.includes(feedback.options.source)) {
```
Comment (dnmeid) [inline analogway_livecore.js:349]: See suggestion above
```suggestion
			if (tallyPGM[feedback.options.source]) {
```

File: analogway_livecore.js [hunk 5]
```diff
@@ -1,743 +1,564 @@
...
-	case 'loadpreset':
-		// set scale
-		if (action.options.scale == '0') {
-			cmd = '0PMlse\n';
-		} else {
-			cmd = '1PMlse\n';
+			return false
 		}
+		if (feedback.type === 'input_previewed') {
+			if (tallyPRV.includes(feedback.options.source)) {
```
Comment (dnmeid) [inline analogway_livecore.js:356]: same as with 'input_active'

File: feedback.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+				bgcolor: this.rgb(200, 0, 0),
+			},
+			// options is how the user can choose the condition the feedback activates for
+			options: [
+				{
+					type: 'number',
+					label: 'Source',
+					id: 'source',
+					default: 1,
+				},
```
Comment (dnmeid) [inline feedback.js:30]: add min and max properties

File: feedback.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+				bgcolor: this.rgb(0, 200, 0),
+			},
+			// options is how the user can choose the condition the feedback activates for
+			options: [
+				{
+					type: 'number',
+					label: 'Source',
+					id: 'source',
+					default: 1,
+				},
```
Comment (dnmeid) [inline feedback.js:51]: add min and max properties

**Review by dnmeid — CHANGES_REQUESTED** [review]
Thanks for giving this old module some love!
See comments in code and please also change package.json

**Comment by kellhogs** [comment]
hey @dnmeid thanks for your sugestions. I will try to implement all of them in the next days.

since im fairly new to the github worklow i have a question. Since ive done some edits scince the upload of the merge request i dont exactly know how to handle the update when i implement your changes. Do i just upload the other edits with the changes or is there a better way to handle this

greetings Fabian

**Comment by kellhogs** [comment]
Forgot about package.json ..... what exactly do you want me to update there?

File: analogway_livecore.js [hunk 8]
```diff
@@ -1,743 +1,574 @@
...
+							this.log('error', 'Received value out of range error from ' + this.config.label + ': ' + line)
+							break
+						default:
+							this.log('error', 'Received unspecified error from Livecore ' + this.config.label + ': ' + line)
+					}
 				}
-			}
 
-		});
+				if (line.match(/TAopr\d+,(0|1)$/)) {
```
Comment (dnmeid) [inline analogway_livecore.js:308]: Still not a Regex which will exclusive be triggered by the TAopr command. Try renaming a memory to TAopr1,1 and see what happens.

**Review by dnmeid — CHANGES_REQUESTED** [review]
It seems you are in the process of a larger modification. As you are the only one currently working at this code, it is perfectly ok to do the development only on your local computer. You can do as many commits in your local repo as you feel you need to. When you are finished and have tested and cleaned  everything and think it is production ready, then you can push your work to your github repo. However on the master branch of the module itself we want to merge PRs as a single commit. That means you should squash your commits into one. If you are working with a lot of commits (which you obviously do) it is also a good idea to implement new stuff in a different branch, that way the master branch stays clean of all the tiny changes of work in progress. Generally please don't do a PR of work in progress, only for finished stuff. If you want to merge different things which you feel are unrelated you can also issue many PRs in different branches, e.g. one for feedbacks, one for presets but you have to make sure that all can be merged without conflicts. Maybe not the best idea for inexperienced people.

Regarding package.json please do a version bump and credit yourself as a contributor.

Regarding the feedbacks in general, you still are missing an initialization of the feedback values after connection to the device. At the moment the feedbacks only reflect the actual device state after a value change has been received. There is a TODO: in the code where the device state should be pulled to initialize the feedback values.

**Comment by kellhogs** [comment]
Thanks for your input @dnmeid. Im new to actively working with github, so i dont know the best practises and best ways do do stuff so your help ist very welcome here. 

i have actually started working in another branch scince im also using github to sync my work between my laptop and my pc.

i've origianally only wanted to implement the feedback for the inputs, therefor the early pull request. From now on i will try to only do major commits in the master branch.

as for the device state for the feedback system: I havent found a way in the documentation to pull all the states from the device, so i would probably need to loop over every single input to get the initial values. I wasn't shure if thats the best way to implement that.

File: package.json [hunk 9]
```diff
@@ -18,6 +18,7 @@
 	"scripts": {
 		"test": "echo \"Error: no test specified\" && exit 1"
 	},
+	"format": "prettier --write .",
```
[the remark this answers is not in the document]
Comment (kellhogs) [reply package.json:21]: Put it in the right place. Documentation was stating that it should be added to the module

**Comment by dnmeid** [comment]
This is now implemented in the latest version, thank you.

**Pull request closed without merging**